DragDrop/src/main/java/io/doist/recyclerviewext/dragdrop/DragDropHelper.java

```diff
@@ -0,0 +1,580 @@
+     * {@link #STATE_RECOVERING}. For any of these except the first, {@link #mViewHolder} is not null.
+     */
+    private int mState = STATE_NONE;
+
+    /**
+     * Dragged item's location on screen.
+     */
+    private Rect mLocation = new Rect();
+    private int mStartLeft;
+    private int mStartTop;
+
+    private int mTouchStartX;
+    private int mTouchStartY;
+    private int mTouchCurrentX;
+    private int mTouchCurrentY;
+
+    private float mScrollSpeed = 0f;
+    private float mScrollSpeedMax;
+
+    /**
+     * Attaches {@link DragDropHelper} to {@code recyclerView}. If already attached to a {@link RecyclerView}, it first
+     * detaches from the previous one. If {@code null} is provided, it detaches from the current {@link RecyclerView}.
+     *
+     * {@link DragDropHelper} uses {@link RecyclerView.ItemDecoration}, {@link RecyclerView.OnItemTouchListener} and
+     * {@link RecyclerView.OnChildAttachStateChangeListener} internally, which are set or removed here.
+     */
+    public void attach(@Nullable RecyclerView recyclerView, @NonNull Callback callback) {
+        if (mRecyclerView != recyclerView) {
+            if (mRecyclerView != null) {
+                mRecyclerView.removeItemDecoration(this);
+                mRecyclerView.removeOnItemTouchListener(this);
+                mRecyclerView.removeOnChildAttachStateChangeListener(this);
+            }
+            mRecyclerView = recyclerView;
+            if (mRecyclerView != null) {
+                mScrollSpeedMax = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, SCROLL_SPEED_MAX_DP,
+                                                            mRecyclerView.getResources().getDisplayMetrics());
+                mRecyclerView.addItemDecoration(this, 0);
+                mRecyclerView.addOnItemTouchListener(this);
+                mRecyclerView.addOnChildAttachStateChangeListener(this);
+            }
+        }
+        mCallback = callback;
+    }
+
+    /**
+     * Starts drag and drop for {@code holder} as soon as there is a touch even, or immediately if one is in progress.
+     *
+     * @return {@code true} if drag is starting, {@code false} if not.
+     */
+    public boolean start(@NonNull RecyclerView.ViewHolder holder) {
+        // Ensure current state is valid and that no other drag is currently in progress.
+        if (holder == mViewHolder) {
+            Log.w(LOG_TAG, "View holder is already being dragged");
+            return false;
+        }
+        if (holder.itemView.getParent() != mRecyclerView) {
+            Log.w(LOG_TAG, "View holder which is not a child of this RecyclerView");
+            return false;
+        }
+        if (holder.getAdapterPosition() == RecyclerView.NO_POSITION) {
+            Log.w(LOG_TAG, "View holder doesn't have an adapter position");
+            return false;
+        }
+        if (mState > STATE_NONE) {
```

Comment:
I'd consider using `!=` instead of `>` here, because it doesn't assume any ordering between the individual states. I don't think we need it, because we only compare `mState` with the the edge state (`STATE_NONE` and `STATE_RECOVERING`) on lines 168, 227, 238 and here, so we can safely replace `>` and `<` with `!=`.